### PR TITLE
Fix C++ compile issue and clang compile issue

### PR DESCRIPTION
**Pull request by x2c3z4**
See commit messages.

### CODE REVIEW

File: src/bthread/mutex.h [hunk 1]
```diff
@@ -48,15 +48,15 @@ class Mutex {
...
-            throw std::system_error(std::error_code(ec, std::system_category()), "Mutex constructor failed");
+            CHECK(false) << "Mutex constructor failed";
         }
     }
     ~Mutex() { CHECK_EQ(0, bthread_mutex_destroy(&_mutex)); }
     native_handler_type native_handler() { return &_mutex; }
     void lock() {
         int ec = bthread_mutex_lock(&_mutex);
         if (ec != 0) {
-            throw std::system_error(std::error_code(ec, std::system_category()), "Mutex lock failed");
```
Comment (lorinlee) [inline src/bthread/mutex.h:59]: Hi @x2c3z4 , Thanks for your contribution! But I have a small question, why remove the exception here, it was been added by #645

File: src/brpc/redis.h [hunk 2]
```diff
@@ -107,21 +107,21 @@ class RedisRequest : public ::google::protobuf::Message {
     bool SerializeTo(butil::IOBuf* buf) const;
 
     // Protobuf methods.
-    RedisRequest* New() const;
-    void CopyFrom(const ::google::protobuf::Message& from);
-    void MergeFrom(const ::google::protobuf::Message& from);
+    RedisRequest* New() const override;
```
Comment (wwbmmm) [inline src/brpc/redis.h:110]: Why "override" is needed only in "redis.h" while not in "thrift_message.h" or "nshead_message.h" or other headers files?

**Comment by wwbmmm** [comment]
This branch has conflicts with master now.

File: src/brpc/redis.h (hunk 2 above)
Comment (x2c3z4) [reply src/brpc/redis.h:110]: Look like the redis.h is included by some very common header, and my project includes the "redis.h" indirectly.

File: src/bthread/mutex.h (hunk 1 above)
Comment (x2c3z4) [reply src/bthread/mutex.h:59]: Google C++ code style suggests that don't use the c++ exception, and I think a lot of projects have disabled the c++ exception including mine. All the brpc code is worked with c++ no-exception, except this code snippet.

If we don't accept this error, just abort it.
As I know, if the mutex has been freed or is a bad memory address, the pthread_mutex_lock will return EINVAL(22 Invalid argument). CHECK(false) is better than c++ exception.

File: src/brpc/redis.h (hunk 2 above)
Comment (wwbmmm) [reply src/brpc/redis.h:110]: I see "redis.h" is included in "server.h".
It is ok to add "override" to "redis.h".

File: src/bthread/mutex.h (hunk 1 above)
Comment (JimChengLin) [reply src/bthread/mutex.h:59]: @x2c3z4 Google Style is good but I guess there is no way to ensure no one else is using this behavior. Removing the exception may break others codebase.

Comment (x2c3z4) [reply src/bthread/mutex.h:59]: What codebase will it break? I don't understand why needs to throw exception here.

**Pull request closed without merging**